### PR TITLE
Fix: #89098 Editor title being read out once suggestion is accepted

### LINKED ISSUE #89098: Editor title being read out once suggestion is accepted
found by @pawelurbanski 

1. Open the suggest widget
2. Accept a suggestion
3. Editor aria label is begin read out :bug:

The issue I believe is that the focus moves back to the editor which makes ScreenReaders read out the editor aria label. I am not sure how to tackle this and am open to suggestions @pawelurbanski 

Assigning to January so I look into fixing this week.
Can reproduce both with NVDA and VoiceOver on mac.

### PR DESCRIPTION
**Pull request by pawelurbanski**
Fixes #89098 with the following changes:
* After suggestion is accepted or canceled aria-label attribute gets removed,
* Switching between opened editors does not change behavior and titles are read out,
* Focusing out of the active editor and back to it gets the title read out,

After suggestion is accepted a screen reader shoudl announce the type of control being: Edit. Contents of the opened file is immediately accessible to scroll up and down to read, so it should be no issue.

I would be happy if @isidorn  could have a look.

### CODE REVIEW

File: src/vs/editor/browser/editorBrowser.ts [hunk 1]
```diff
@@ -332,7 +332,8 @@ export interface IOverviewRuler {
  * @internal
  */
 export interface IEditorAriaOptions {
-	activeDescendant: string | undefined;
+	activeDescendant?: string | undefined;
```
Comment (isidorn) [inline src/vs/editor/browser/editorBrowser.ts]: I do not like that these attributes are both optional and can be `undefined`.
I would just go with one approach and stick with that. Either they are optional or can be `undefined`

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@pawelurbanski thanks a lot for this PR.
I have commented directly in the PR and potentially requested some changed.
Great work!

**Comment by pawelurbanski** [comment]
The idea is as follows:
When the following takes place: suggestion accepted || selection canceled then editor gets back focus.
When it happens the screen reader reads the name / accessible name assigned by the aria-label and later on starts reading the current line.
What we want to get is:
Have the title of the editor set initially, but clear it once either of 2 events mentioned above takes place, but before focus gets back to the editor textarea.
The key is to have the label set to nothing when focus gets back to the editor. If we can do it in the other place as you mentioned and get this effect then I’m fine. Do keep in mind that suggestion completions are triggered by the Tab or Enter keys and when escaping by hitting the Escape key the label must be clear or we get thic chatty behavior we want to avoid in the first place.
As for the code:
I made activeDescendant and ariaLabel optional because when making a call to setAriaOptions on the editor I was required to provide the element ID of the activeDescendant. I was not that familiar with the widget code and took that step, keeping in mind that it would require further work.

**Comment by isidorn** [comment]
@pawelurbanski great, the idea makes perfect sense and I really like the approach.
However I believe it will also work if we do it in the other place. Let me know if you want to change the code or you would like me to update it?
Also did you test this PR out that it works as expected?

**Comment by pawelurbanski** [comment]
I will test if doing the change in a place where you manage the state for activeDescendant works as well. I tested the change running the dev build with and without my screen reader plugin.
Github build fails and it needs investigation. I run this task locally and this error was not returned. I run the mentioned action on my local brunch, but there seemed to be no changes in the file.

**Comment by isidorn** [comment]
@pawelurbanski GitHub build is failing due to unrelated changes (some other files needs formating). You can ignore it for now and I can take care of it.
Once you try it out let me know and we can look into merging this. Thanks

**Comment by pawelurbanski** [comment]
I changed the place where setARIAOptions method is called and removes the aria-label from the textarea input field, which is the editor.
The only remaining issue is how to switch back to treating ariaLabel and activeDescendant as required.
After investigating a bit how NVDA for example behaves we deal with the following case in rought order. It would mean that the aria-label attribute would always be removed after any completion that triggers: event_loseFocus in NVDA for example. Once the suggestion is inserted the editor textarea is still unfocused and gets focus back on the first character/key pressed. This character causes change of editor contents, which triggers the suggestion widget as part of the event loop. Once we set aria-activedescendant on the next suggestion, we still want to keep the aria-label removed for we are editing the same file and do not want the title to be read out.
I still need to figure out why once the suggestion wiggets pops up and the editor focused state is false the "has popup" and "autocomplete" states are being read by the screen reader. It does not happen when using the development build, but happens in stable release or insider build as of this writing.

File: src/vs/editor/contrib/suggest/suggestWidget.ts [hunk 2]
```diff
@@ -751,6 +751,7 @@ export class SuggestWidget implements IContentWidget, IListVirtualDelegate<Compl
 			}
 
 			this.editor.setAriaOptions({ activeDescendant: undefined });
+			this.editor.setAriaOptions({ ariaLabel: undefined });
```
Comment (isidorn) [inline src/vs/editor/contrib/suggest/suggestWidget.ts]: I would just merge these two calls into one.

File: src/vs/editor/browser/controller/textAreaHandler.ts [hunk 3]
```diff
@@ -123,6 +123,7 @@ export class TextAreaHandler extends ViewPart {
 		this.textArea.setAttribute('autocapitalize', 'off');
 		this.textArea.setAttribute('autocomplete', 'off');
 		this.textArea.setAttribute('spellcheck', 'false');
+		this.textArea.setAttribute('autofocus', 'true');
```
Comment (isidorn) [inline src/vs/editor/browser/controller/textAreaHandler.ts]: Why the `autofocus`?

**Comment by isidorn** [comment]
@pawelurbanski thanks for providing details. I have added two minor comments in the changes.
Next week we are wrapping up the milestone, so if we would like to include this as part of the release we should wrap up until Wednesday. I can also jump on it or we can also do it for March with no hurries. Whatever works for you.

Adding @jrieken as reviewer since he owns the suggest widget.

**Review by jrieken — APPROVED** [review]
Looking good for suggest

**Comment by pawelurbanski** [comment]
I added autofocus since in theory it hints the browser to place focus immediately in that element. VS Code can have other editable fields in the view, so that in theory would get focus to the main editor area.
I will merge the call to the setARIAOptions function and push the change to the branch. It would be great to get this fix out the door and maybe test the results in the insider build before the release.

**Comment by isidorn** [comment]
@pawelurbanski we usualy manage focus on our own, and if that `autofocus` does not bring anything to the screen reader I would remove it. And I think it does not.
Yeah, we can try to merge this in today or Tuesday. Please note that I am on vacation on Monday.

**Comment by pawelurbanski** [comment]
All done. It seems that the autofocus attribute was redundant so I removed it. Set back the activeDescendant parameter to required and combined the call.

**Comment by isidorn** [comment]
@pawelurbanski Thanks a lot. I have tried this PR out. And the issues I see:
1) Once the suggestion is accepted even though the editor's `ariaLabel` is empty now "edit  has auto complete  multi line" is still being read out, since we focus moves to the editor. I am not sure if we can fix this. Do you see the same beahvior
2) We never reset the editor's `ariaLabel`. I believe we can only do that with some timeout solution. The problem with the current approach: accept a suggestion, move focus to somewhere else, bring focus back to editor -> the editor title is not read
3) We are still setting the `autoFocus` in the PR. I think you forgot to remove this
4) There is an unused static `NO_SUGGESTIONS_MESSAGE`. We should remove it.

**Comment by pawelurbanski** [comment]
@isidorn Here is a bit of simple pseudodebugging to illustrate the interaction of NVDA with the editor and suggest widget:
I’m posting the output from the speech history window with some help of my add-on. The lines with words: Lost and Gain mean that the editor lost focus or got it back:
markersView.ts editor


Gain
t
has auto complete  multi line
Lost
this
markersView.ts editor
                                               this

Gain
dot
has auto complete  multi line
Lost
_expanded
e
markersView.ts editor
                                               this.e

Gain
has auto complete  multi line
Lost
editorService
d
editorService
markersView.ts editor
                                               this.editorService

Gain
It looks to me that the flow is as follows:

  *   Focus is in the editor,
  *   A character is typed,
  *   Thre is a suggestion match,
  *   We set the aria-descendant,
  *   Change attributes on the textarea tag in setAriaOptions,
  *   Screen processes so called positive states change and reports it back,
  *   Editor loses focus and the suggest item is presented,
  *   Focus is still outside when you keep typing, but suggestion matches,
  *   Focus is brought back when match changes,
  *   See the flow from suggestion “_expanded” to “e”.

What I do now is I drop edit control states and rename the edit role to nothing. The “edit” read out before the editor loses focus is the role of the control.

I wonder if we could change the attributes for autocomplete and so on after the editor loses focus, namely after the suggestion widget is activated and we pass the aria-activeDescendant. But I guess, it would break the HTML and browser mechanism of using those attributes and autocomplete set to “list” or maybe not.

As for other fixes – I will fix them. Apologies for this ommition and causing you unnecessary work.

**Comment by isidorn** [comment]
Thanks for providing more feedback. Yes that seems to be happening, and I do agree we need to somehow hack it in that when the editor "gets focus" such that not too much is read out. We can aslo ask on NVDA github page for ideas how to potentially achieve this.
Also if we move paramter hints to also use `activeDescendetes` we will have the same issue, once the paremeterHints get closed the editor aria label we get re-read.

A crazy idea which I have and I am not sure if it will work is that we set `ariaHidden=true` on the editor instead of cleaning out its ariaLabel. The focus goes back to the editor, nothing gets read out and after a timeout we remove this `ariaHidden` attribute. That would sort of make the editor invisible to the screen reader for a very brief perioud of time.

For other fixes: no problem, no need for appologies :)

**Comment by pawelurbanski** [comment]
@isidorn I have the idea for a fix with NVDA. Truth be told I read through the source code to figure out why this reading may happen. Since I need a good IDE for some types of my work with are mostly accessibility unrelated, I can still make this better.

**Comment by isidorn** [comment]
@pawelurbanski even if you fix this on the NVDA side users using Orca on Linux and VoiceOver on Mac will still hit this issue. I would still explore with the `ariaHidden` idea since that might benefit all. I can also try it out some time next week.

**Comment by pawelurbanski** [comment]
Since the completion widget responds to keyboard events and we can scroll up/down through the items. Why not try and just set focus to the currently active item returned from the completion provider?
Theoretically, we would switch focus back and forth between the completion list and the editor.

As for screen readers – their default behavior is optimized for different cases. When you are casually browsing the web hearing this short control type label: “edit” every now and then, even with the longer states labels is not an issue.
Our case is much different though. We switch focus a few times a second if you are afast typer and intelisense can keep up.
To my knowledge: Orca does allow for making scripts. But it is not the mainstream user case when it comes to OS and screen reader combination. Voiceover does allow for scripts written in AppleScript, but it seems that they are not as advanced as the ones provided by NVDA or JAWS – no ability to inspect object / control properties I guess.
Last thing to add is that VS Code is a tool targeting at least the power user, who can take the effort to install an NVDA add-on or a JAWS script.

**Comment by isidorn** [comment]
All good points.
However I would like to experiment with the ariaHidden idea, just to see if it makes any sense before we proceed here.
Also I was on a short vacation, and thus my slow response.
Bottom line -> moving this to next milestone so we have time then to experiment more with this. I plan to continue on this start of next week.
Thanks a lot for the help

**Comment by isidorn** [comment]
Closing this as we do not plan to merge this in

**Pull request closed without merging**